src/coord/src/coord.rs

```diff
@@ -5182,54 +5183,68 @@ mod timeline {
     /// visible to exactly reads at that time or greater, and no other times.
-    enum TimestampOracleState {
+    enum TimestampOracleState<T> {
         /// The timeline is producing collection updates timestamped with the argument.
-        Writing(mz_repr::Timestamp),
+        Writing(T),
         /// The timeline is observing collections aot the time of the argument.
-        Reading(mz_repr::Timestamp),
+        Reading(T),
     }
 
     /// A type that provides write and read timestamps, reads observe exactly their preceding writes..
     ///
     /// Specifically, all read timestamps will be greater or equal to all previously reported write timestamps,
     /// and strictly less than all subsequently emitted write timestamps.
-    pub struct TimestampOracle {
-        state: TimestampOracleState,
-        advance_to: Option<mz_repr::Timestamp>,
+    pub struct TimestampOracle<T> {
+        state: TimestampOracleState<T>,
+        advance_to: Option<T>,
+        next: Box<dyn Fn() -> T>,
     }
 
-    impl TimestampOracle {
-        /// Create a new timeline, starting at the indicated time.
-        pub fn new(initially: mz_repr::Timestamp) -> Self {
+    impl<T: super::CoordTimestamp> TimestampOracle<T> {
+        /// Create a new timeline, starting at the indicated time. `next` generates
+        /// new timestamps when invoked. The timestamps have no requirements, and can
+        /// retreat from previous invocations.
+        pub fn new<F>(initially: T, next: F) -> Self
+        where
+            F: Fn() -> T + 'static,
+        {
             Self {
-                state: TimestampOracleState::Writing(initially),
+                state: TimestampOracleState::Writing(initially.clone()),
                 advance_to: Some(initially),
+                next: Box::new(next),
             }
         }
 
         /// Acquire a new timestamp for writing.
         ///
-        /// This timestamp will be strictly greater than all prior values of `self.read_ts()`,
-        /// and less than or equal to all subsequent values of `self.read_ts()`.
-        pub fn write_ts(&mut self) -> mz_repr::Timestamp {
-            match self.state {
-                TimestampOracleState::Writing(ts) => ts,
+        /// This timestamp will be strictly greater than all prior values of
+        /// `self.read_ts()`, and less than or equal to all subsequent values of
+        /// `self.read_ts()`.
+        pub fn write_ts(&mut self) -> T {
+            match &self.state {
+                TimestampOracleState::Writing(ts) => ts.clone(),
                 TimestampOracleState::Reading(ts) => {
-                    self.state = TimestampOracleState::Writing(ts + 1);
-                    ts + 1
+                    let mut next = (self.next)();
+                    if next.less_equal(&ts) {
+                        next = ts.step_forward();
+                    }
+                    self.state = TimestampOracleState::Writing(next.clone());
```

Comment:
Nit, but in light of the `saturating_add` question, maybe an `assert!(ts.less_than(next))` here?